src/extension.ts

```diff
@@ -36,23 +38,31 @@ export function activate(context: ExtensionContext) {
             diagnostic.source
               ? has(["ts", "deno-ts", "js"], diagnostic.source)
               : false
-          )
-          .forEach(async (diagnostic) => {
+          ).forEach(async (diagnostic) => {
             // formatDiagnostic converts message based on LSP Diagnostic type, not VSCode Diagnostic type, so it can be used in other IDEs.
             // Here we convert VSCode Diagnostic to LSP Diagnostic to make formatDiagnostic recognize it.
-            const markdownString = new MarkdownString(
-              formatDiagnostic(converter.asDiagnostic(diagnostic), prettify)
-            );
+            let formattedMessage = cache.get(diagnostic.message);
 
-            markdownString.isTrusted = true;
-            markdownString.supportHtml = true;
+            if (!formattedMessage) {
+              // All the expensive formatting is happening here in formatDiagnosticMessage
+              const markdownString = new MarkdownString(
+                formatDiagnostic(converter.asDiagnostic(diagnostic), prettify)
+              );
+              
+              markdownString.isTrusted = true;
+              markdownString.supportHtml = true;
+              
+              formattedMessage = {
+                range: diagnostic.range,
+                contents: [markdownString],
+              };
```

Comment:
We need to save only the formatted message in the cache without the range:

```typescript
formattedMessage = markdownString;
cache.set(diagnostic.message, formattedMessage);
```
and then in push we'll add the range:
```typscript
items.push({
 range: diagnostic.range,
 contents: [formattedMessage],
});
```

That way, the message will not appear twice or in the wrong places 🙂